fix: make sure the destination does not exist before organizing

upils opened the pull request:
- [x] Have you followed the guidelines for contributing?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `make lint && make test`?
- [x] Have you added an entry to the changelog (`docs/reference/changelog.rst`)?

---

Fixes #1366
CRAFT-4870

jahn-junior commented on docs/reference/changelog.rst:
```diff
@@ -27,6 +27,10 @@ Bug fixes:
+- `#1366 <https://github.com/canonical/craft-parts/issues/1366>`_ When the destination
+  of an ``organize`` entry is a directory, properly detect collisions with existing
+  files in this directory. A proper error is now returned, giving necessary information
+  to the user to resolve the conflict.
```
```suggestion
- `#1366 <https://github.com/canonical/craft-parts/issues/1366>`_ When organizing a file
  to a directory, detect collisions with existing files in the directory. If a collision
  occurs, an error is produced with a message listing the colliding paths.
```

upils commented:
> This looks good in general, but there's a part of me that wonders whether it could cause re-run issues. Do we have any integration tests that would cover that case?

@lengau in case of a rerun the build step is either skipped, updated or cleaned and re-executed.

If it is updated, the `overwrite` param is set to true and the previous branch (just before my additional change) is executed, removing the existing file. https://github.com/canonical/craft-parts/blob/4c4c228d7d65f1d712b3b80e502f86f91201d5b4/tests/unit/features/partitions/executor/test_organize.py#L203 is testing that.

If it is cleaned, the existing file is also removed.

So I do not see a case where it could prevent from reruning.
